nixos/modules/services/editors/haste.nix

```diff
@@ -0,0 +1,86 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  pkg = pkgs.haste-server;
+  cfg = config.services.haste-server;
+
+  format = pkgs.formats.json {};
+in
+{
+  options.services.haste-server = {
+    enable = mkEnableOption "haste-server";
+    openFirewall = mkEnableOption "firewall passthrough for haste-server";
+
+    settings = mkOption {
+      description = ''
+        Configuration for haste-server.
+        For documentation see <link xlink:href="https://github.com/toptal/haste-server#settings">project readme</link>
+      '';
+      type = format.type;
+    };
+  };
+
+  config = mkIf (cfg.enable) {
+    networking.firewall.allowedTCPPorts = mkIf (cfg.openFirewall) [ cfg.settings.port ];
+
+    services.haste-server = {
+      settings = mkDefault {
+        host = "::";
+        port = 7777;
+
+        keyLength = 10;
+        maxLength = 400000;
+
+        staticMaxAge = 86400;
+        recompressStaticAssets = false;
+
+        logging = [
+          {
+            level = "verbose";
+            type = "Console";
+            colorize = true;
+          }
+        ];
+
+        keyGenerator = {
+          type = "phonetic";
+        };
+
+        rateLimits = {
+          categories = {
+            normal = {
+              totalRequests = 500;
+              every = 60000;
+            };
+          };
+        };
```

Comment:
Will something like this merge how you want them to if the user provides other key-values here?

Reply:
ah crap, doesn't work. any way I can fix that?

Reply:
You probably want something similar to this:

https://github.com/NixOS/nixpkgs/blob/47cd6702934434dd02bc53a67dbce3e5493e33a2/nixos/modules/services/mail/davmail.nix#L61-L81